lib/cisco_node_utils/span_session.rb

```diff
@@ -0,0 +1,123 @@
+# Insert appropriate license here
+require_relative 'node_util'
+require_relative 'interface'
+
+module Cisco
+  # node utils class for SPAN (switchport analyzer) sessions
+  class SpanSession < NodeUtil
+    attr_reader :session_id
+
+    def initialize(session_id, instantiate=true)
+      validate_args(session_id.to_i)
+      create if instantiate
+    end
+
+    def self.sessions
+      hash = {}
+      all = config_get('span_session', 'all_sessions')
+      return hash if all.nil?
+
+      all.each do |id|
+        hash[id] = SpanSession.new(id, false)
+      end
+      hash
+    end
+
+    def validate_args(session_id)
+      fail TypeError unless session_id.is_a?(Integer)
+      @session_id = session_id
+    end
+
+    def create
+      config_set('span_session', 'create', id: @session_id)
+    end
+
+    def destroy
+      config_set('span_session', 'destroy', id: @session_id)
+    end
+
+    def description
+      config_get('span_session', 'description', id: @session_id)
+    end
+
+    def description=(val)
+      val = val.to_s
+      if val.empty?
+        config_set('span_session', 'description', id: @session_id,
+                    state: 'no', description: '')
+      else
+        config_set('span_session', 'description', id: @session_id,
+                    state: '', description: val)
+      end
+    end
+
+    def destination
+      config_get('span_session', 'destination', id: @session_id)
+    end
+
+    def destination=(int)
+      # fail if int is not a valid interface
+      fail TypeError unless Interface.interfaces.key?(int)
+      config_set('span_session', 'destination', id: @session_id, intf_name: int)
+    end
+
+    def session_id
+      config_get('span_session', 'session_id')
+    end
+
+    def session_id=(id)
+      fail TypeError unless id.is_a?(Integer)
+      config_set('span_session', 'session_id', id: id, state: '')
+    end
+
+    def shutdown
+      config_get('span_session', 'shutdown', id: @session_id)
+    end
+
+    def shutdown=(bool)
+      fail TypeError unless bool.is_a?(Boolean)
+      config_set('span_session', 'shutdown', id: @session_id, shutdown: bool)
+    end
+
+    def source_interface
+      config_get('span_session', 'source_interface', id: @session_id)
+    end
+
+    def source_interface=(sources)
```

Comment:
So for the source interfaces/vlans, I was hoping to use a Hash only because the data is both the interface names/vlan ID's in addition to the direction. I know the `cisco_bgp_af` uses an array-of-arrays for the `networks` property but this was awkward to use with Puppet DSL w/ Hiera data lookups and required a custom function in Ruby to transform a basic Puppet-native data type into the one the property was expecting. There's no clean and easy way to take in (for example) a hash via Hiera then transform it into an array of array pairs. 

Thoughts?
Ideally, my end result as a user of ciscopuppet would be something like this:
```
$span_sources = { 'Ethernet1/1' => 'both', 'Ethernet1/2', => 'in', 'portchannel-10' => 'out' }

cisco_span_session { '5':
  ensure => present,
  description => 'local_span_session',
  shutdown => false,
  type => 'local',
  source_interfaces => $span_sources,
  destination_interface => 'Ethernet1/48'
}
```